// @aws-crypto/kms-keyring-node/build/main/kms_keyring_node.js
"use strict";
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
const kms_keyring_1 = require("@aws-crypto/kms-keyring");
exports.getClient = kms_keyring_1.getClient;
exports.limitRegions = kms_keyring_1.limitRegions;
exports.excludeRegions = kms_keyring_1.excludeRegions;
exports.cacheClients = kms_keyring_1.cacheClients;
const material_management_node_1 = require("@aws-crypto/material-management-node");
const aws_sdk_1 = require("aws-sdk");
exports.KMS = aws_sdk_1.KMS;
const getKmsClient = kms_keyring_1.getClient(aws_sdk_1.KMS, {
    customUserAgent: 'AwsEncryptionSdkJavascriptNodejs',
});
exports.getKmsClient = getKmsClient;
const cacheKmsClients = kms_keyring_1.cacheClients(getKmsClient);
exports.cacheKmsClients = cacheKmsClients;
class KmsKeyringNode extends kms_keyring_1.KmsKeyringClass(material_management_node_1.KeyringNode) {
    constructor({ clientProvider = cacheKmsClients, keyIds, generatorKeyId, grantTokens, discovery, } = {}) {
        super({ clientProvider, keyIds, generatorKeyId, grantTokens, discovery });
    }
}
exports.KmsKeyringNode = KmsKeyringNode;
material_management_node_1.immutableClass(KmsKeyringNode);
