// @aws-crypto/encrypt-node/build/main/encrypt_stream.js
"use strict";
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const material_management_node_1 = require("@aws-crypto/material-management-node");
const framed_encrypt_stream_1 = require("./framed_encrypt_stream");
const signature_stream_1 = require("./signature_stream");
const duplexify_1 = __importDefault(require("duplexify"));
const crypto_1 = require("crypto");
const serialize_1 = require("@aws-crypto/serialize");
// @ts-ignore
const readable_stream_1 = require("readable-stream");
const fromUtf8 = (input) => Buffer.from(input, 'utf8');
const { serializeMessageHeader, headerAuthIv } = serialize_1.serializeFactory(fromUtf8);
/**
 * Takes a NodeDefaultCryptographicMaterialsManager or a KeyringNode that will
 * be wrapped in a NodeDefaultCryptographicMaterialsManager and returns a stream.
 *
 * @param cmm NodeMaterialsManager|KeyringNode
 * @param op EncryptStreamInput
 */
function encryptStream(cmm, op = {}) {
    const { suiteId, encryptionContext = {}, frameLength = serialize_1.FRAME_LENGTH, plaintextLength, } = op;
    /* Precondition: The frameLength must be less than the maximum frame size Node.js stream. */
    material_management_node_1.needs(frameLength > 0 && serialize_1.Maximum.FRAME_SIZE >= frameLength, `frameLength out of bounds: 0 > frameLength >= ${serialize_1.Maximum.FRAME_SIZE}`);
    /* If the cmm is a Keyring, wrap it with NodeDefaultCryptographicMaterialsManager. */
    cmm =
        cmm instanceof material_management_node_1.KeyringNode
            ? new material_management_node_1.NodeDefaultCryptographicMaterialsManager(cmm)
            : cmm;
    const suite = suiteId && new material_management_node_1.NodeAlgorithmSuite(suiteId);
    const wrappingStream = new duplexify_1.default();
    cmm
        .getEncryptionMaterials({ suite, encryptionContext, plaintextLength })
        .then(async (material) => {
        const { dispose, getSigner } = material_management_node_1.getEncryptHelper(material);
        const { getCipher, messageHeader, rawHeader } = getEncryptionInfo(material, frameLength);
        wrappingStream.emit('MessageHeader', messageHeader);
        const encryptStream = framed_encrypt_stream_1.getFramedEncryptStream(getCipher, messageHeader, dispose, plaintextLength);
        const signatureStream = new signature_stream_1.SignatureStream(getSigner);
        readable_stream_1.pipeline(encryptStream, signatureStream);
        wrappingStream.setReadable(signatureStream);
        // Flush the rawHeader through the signatureStream
        rawHeader.forEach((buff) => signatureStream.write(buff));
        // @ts-ignore until readable-stream exports v3 types...
        wrappingStream.setWritable(encryptStream);
    })
        .catch((err) => wrappingStream.emit('error', err));
    return wrappingStream;
}
exports.encryptStream = encryptStream;
function getEncryptionInfo(material, frameLength) {
    const { kdfGetCipher } = material_management_node_1.getEncryptHelper(material);
    const { encryptionContext } = material;
    const messageId = crypto_1.randomBytes(serialize_1.MESSAGE_ID_LENGTH);
    const { id, ivLength } = material.suite;
    const messageHeader = Object.freeze({
        version: serialize_1.SerializationVersion.V1,
        type: serialize_1.ObjectType.CUSTOMER_AE_DATA,
        suiteId: id,
        messageId,
        encryptionContext,
        encryptedDataKeys: Object.freeze(material.encryptedDataKeys),
        contentType: serialize_1.ContentType.FRAMED_DATA,
        headerIvLength: ivLength,
        frameLength,
    });
    const { buffer, byteOffset, byteLength } = serializeMessageHeader(messageHeader);
    const headerBuffer = Buffer.from(buffer, byteOffset, byteLength);
    const info = serialize_1.kdfInfo(messageHeader.suiteId, messageHeader.messageId);
    const getCipher = kdfGetCipher(info);
    const headerIv = headerAuthIv(ivLength);
    const validateHeader = getCipher(headerIv);
    validateHeader.setAAD(headerBuffer);
    validateHeader.update(Buffer.alloc(0));
    validateHeader.final();
    const headerAuth = validateHeader.getAuthTag();
    return {
        getCipher,
        messageHeader,
        rawHeader: [headerBuffer, headerIv, headerAuth],
    };
}
exports.getEncryptionInfo = getEncryptionInfo;
