// @aws-crypto/raw-keyring/build/module/raw_aes_material.js
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
/* Here I am reusing the Material implementation and interface from material-management.
 * This is because there are many security guarantees that this implementations offer
 * that map to the current implementation of raw AES keyrings.
 * The KeyringTrace is an unfortunate case because there is no mapping.
 * However the trade off seems worth it and the convolutions to make it work seem minimal.
 */
import { decorateCryptographicMaterial, decorateWebCryptoMaterial, frozenClass, NodeAlgorithmSuite, WebCryptoAlgorithmSuite, KeyringTraceFlag, needs, } from '@aws-crypto/material-management';
import { RawAesWrappingSuiteIdentifier, } from './raw_aes_algorithm_suite';
export class NodeRawAesMaterial {
    constructor(suiteId) {
        this.keyringTrace = [];
        this.encryptionContext = Object.freeze({});
        /* Precondition: NodeRawAesMaterial suiteId must be RawAesWrappingSuiteIdentifier. */
        needs(RawAesWrappingSuiteIdentifier[suiteId], 'suiteId not supported.');
        this.suite = new NodeAlgorithmSuite(suiteId);
        /* NodeRawAesMaterial need to set a flag, this is an abuse of TraceFlags
         * because the material is not generated.
         * but CryptographicMaterial force a flag to be set.
         */
        const setFlags = KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY;
        decorateCryptographicMaterial(this, setFlags);
        Object.setPrototypeOf(this, NodeRawAesMaterial.prototype);
        Object.freeze(this);
    }
    hasValidKey() {
        return this.hasUnencryptedDataKey;
    }
}
frozenClass(NodeRawAesMaterial);
export class WebCryptoRawAesMaterial {
    constructor(suiteId) {
        this.keyringTrace = [];
        this.encryptionContext = Object.freeze({});
        /* Precondition: WebCryptoAlgorithmSuite suiteId must be RawAesWrappingSuiteIdentifier. */
        needs(RawAesWrappingSuiteIdentifier[suiteId], 'suiteId not supported.');
        this.suite = new WebCryptoAlgorithmSuite(suiteId);
        this.validUsages = Object.freeze([
            'decrypt',
            'encrypt',
        ]);
        /* WebCryptoRawAesMaterial need to set a flag, this is an abuse of TraceFlags
         * because the material is not generated.
         * but CryptographicMaterial force a flag to be set.
         */
        const setFlag = KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY;
        decorateCryptographicMaterial(this, setFlag);
        decorateWebCryptoMaterial(this, setFlag);
        Object.setPrototypeOf(this, WebCryptoRawAesMaterial.prototype);
        Object.freeze(this);
    }
    hasValidKey() {
        return this.hasUnencryptedDataKey && this.hasCryptoKey;
    }
}
frozenClass(WebCryptoRawAesMaterial);
