// @aws-crypto/encrypt-node/build/module/encrypt_stream.js
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { NodeDefaultCryptographicMaterialsManager, NodeAlgorithmSuite, KeyringNode, getEncryptHelper, needs, } from '@aws-crypto/material-management-node';
import { getFramedEncryptStream } from './framed_encrypt_stream';
import { SignatureStream } from './signature_stream';
import Duplexify from 'duplexify';
import { randomBytes } from 'crypto';
import { serializeFactory, kdfInfo, ContentType, SerializationVersion, ObjectType, FRAME_LENGTH, MESSAGE_ID_LENGTH, Maximum, } from '@aws-crypto/serialize';
// @ts-ignore
import { pipeline } from 'readable-stream';
const fromUtf8 = (input) => Buffer.from(input, 'utf8');
const { serializeMessageHeader, headerAuthIv } = serializeFactory(fromUtf8);
/**
 * Takes a NodeDefaultCryptographicMaterialsManager or a KeyringNode that will
 * be wrapped in a NodeDefaultCryptographicMaterialsManager and returns a stream.
 *
 * @param cmm NodeMaterialsManager|KeyringNode
 * @param op EncryptStreamInput
 */
export function encryptStream(cmm, op = {}) {
    const { suiteId, encryptionContext = {}, frameLength = FRAME_LENGTH, plaintextLength, } = op;
    /* Precondition: The frameLength must be less than the maximum frame size Node.js stream. */
    needs(frameLength > 0 && Maximum.FRAME_SIZE >= frameLength, `frameLength out of bounds: 0 > frameLength >= ${Maximum.FRAME_SIZE}`);
    /* If the cmm is a Keyring, wrap it with NodeDefaultCryptographicMaterialsManager. */
    cmm =
        cmm instanceof KeyringNode
            ? new NodeDefaultCryptographicMaterialsManager(cmm)
            : cmm;
    const suite = suiteId && new NodeAlgorithmSuite(suiteId);
    const wrappingStream = new Duplexify();
    cmm
        .getEncryptionMaterials({ suite, encryptionContext, plaintextLength })
        .then(async (material) => {
        const { dispose, getSigner } = getEncryptHelper(material);
        const { getCipher, messageHeader, rawHeader } = getEncryptionInfo(material, frameLength);
        wrappingStream.emit('MessageHeader', messageHeader);
        const encryptStream = getFramedEncryptStream(getCipher, messageHeader, dispose, plaintextLength);
        const signatureStream = new SignatureStream(getSigner);
        pipeline(encryptStream, signatureStream);
        wrappingStream.setReadable(signatureStream);
        // Flush the rawHeader through the signatureStream
        rawHeader.forEach((buff) => signatureStream.write(buff));
        // @ts-ignore until readable-stream exports v3 types...
        wrappingStream.setWritable(encryptStream);
    })
        .catch((err) => wrappingStream.emit('error', err));
    return wrappingStream;
}
export function getEncryptionInfo(material, frameLength) {
    const { kdfGetCipher } = getEncryptHelper(material);
    const { encryptionContext } = material;
    const messageId = randomBytes(MESSAGE_ID_LENGTH);
    const { id, ivLength } = material.suite;
    const messageHeader = Object.freeze({
        version: SerializationVersion.V1,
        type: ObjectType.CUSTOMER_AE_DATA,
        suiteId: id,
        messageId,
        encryptionContext,
        encryptedDataKeys: Object.freeze(material.encryptedDataKeys),
        contentType: ContentType.FRAMED_DATA,
        headerIvLength: ivLength,
        frameLength,
    });
    const { buffer, byteOffset, byteLength } = serializeMessageHeader(messageHeader);
    const headerBuffer = Buffer.from(buffer, byteOffset, byteLength);
    const info = kdfInfo(messageHeader.suiteId, messageHeader.messageId);
    const getCipher = kdfGetCipher(info);
    const headerIv = headerAuthIv(ivLength);
    const validateHeader = getCipher(headerIv);
    validateHeader.setAAD(headerBuffer);
    validateHeader.update(Buffer.alloc(0));
    validateHeader.final();
    const headerAuth = validateHeader.getAuthTag();
    return {
        getCipher,
        messageHeader,
        rawHeader: [headerBuffer, headerIv, headerAuth],
    };
}
